### ISSUE TITLE
How to log important dropped spans errors in AsyncReporter?

### ISSUE THREAD

**Issue by dzou** [openzipkin/zipkin-reporter-java#154]
We have a problem where sometimes all spans will be dropped in Stackdriver because 
the user might not have configured their GCP project correctly. The exception thrown will give the user instructions for how to fix it, however [AsyncReporter will log these errors at debug level on this line](https://github.com/openzipkin/zipkin-reporter-java/blob/master/core/src/main/java/zipkin2/reporter/AsyncReporter.java#L289). We had one user get blocked on this for a long time because they did not see the error; they only found it after changing the debug level to DEBUG by chance.

context: https://github.com/spring-cloud/spring-cloud-gcp/issues/1752

Is there some way to configure or override this behavior to specify that certain special errors are "fatal" or from user error and should be logged?

In this case, a `StatusRuntimeException` was thrown (this is a custom GRPC exception). We would like to be able to check the error code of this exception and if it is considered 'fatal', be able to log the error at error-level.

Is something like this possible?

**Comment by adriancole** [comment]
another option is to invoke reporter/sender.check() eagerly or once and log severely if it fails.

We do something similar to this for config problems in armeria.

https://github.com/line/armeria/blob/master/brave/src/main/java/com/linecorp/armeria/common/brave/RequestContextCurrentTraceContext.java#L117-L134

**Comment by adriancole** [comment]
note: some configuration is out of scope for this level of abstraction. For example, in Zipkin-GCP some configuration is verified eagerly, for example OpenSSL setup which could be an option for truly configuration related setup https://github.com/openzipkin/zipkin-gcp/blob/fd98d07e1bdeddab21e1484c7ae48631d9e824d0/autoconfigure/storage-stackdriver/src/main/java/zipkin/autoconfigure/storage/stackdriver/ZipkinStackdriverStorageAutoConfiguration.java#L92-L96

**Comment by adriancole** [comment]
sounded an alarm also on our dev list here https://groups.google.com/forum/#!topic/zipkin-dev/4Bo9e4ETLNc

**Comment by bbzg** [comment]
I sympathize with your goal of not locking up peoples cloud consoles, but I personally (a lowly 4-cpu cloud user) have found two issues in this space in a short amount of time, that I can only guess is affecting a lot of users.

(1) The first one I found was this: https://github.com/spring-cloud/spring-cloud-gcp/issues/1705 a misconfiguration of spring-cloud-gcp that caused an extra unused reporter that constantly failed without error output (wasting a bit of cpu every time).

(2) The second one, kindly cross-reported by @dzou is the one in this thread.

And all I had to do was to spend quite a few hours of my time stepping into the code of spring-cloud-gcp and zipkin to try to figure out which logs to even turn on...

So if this could be solved in a way that caters to both the developers of e.g. spring-cloud-gcp by surfacing the types of errors that I found in (1) by making sure they see them while they are doing their library development, and by surfacing the types of errors in (2) so that I can see them while doing application development - that would be great.

Thank you for all your great work creating zipkin!

**Comment by adriancole** [comment]
@bbzg thanks for the feedback. Both of the scenarios mentions are config related and should be something that could be surfaced well before use. For example, calling .check() should have resulted in the permissions error in configuration. The duplication of reporter beans is way out of scope for this repo. My suggestion would be to enlist the existing health check function for this kind of problem as there's no reason to wait to do that. As doing that is framework specific, it should be done there (ex boot and dropwizard will have different mechanisms for post-construct tasks that should happen prior to use, but not during constructors).

ps while we didn't create zipkin, we did create all the code currently released. anyway thanks for the thanks :P

**Comment by adriancole** [comment]
I added an item here about configuration problems. https://github.com/spring-cloud/spring-cloud-sleuth/issues/1411

Let's reserve discussion on this thread to things that cannot be caught at configure time.

**Comment by adriancole** [comment]
PS I'm kindof backlogged on other things at the moment, but I do think we could do something here.. probably a low frequency log message (ex once in 5 minutes) to severe level saying check metrics or debug logs for more. we do something similar in our cassandra code I think (cc @michaelsembwever). a low frequency message would be slightly better than the slf4j style log once at startup because it can also show runtime problems.

**Comment by shakuzen** [comment]
> And all I had to do was to spend quite a few hours of my time stepping into the code of spring-cloud-gcp and zipkin to try to figure out which logs to even turn on...

Most likely not the whole solution, but this makes me think that we should document what logger to raise when debugging this type of issue, so people don't have to dig in the code. @adriancole's suggestion of a log mentioning this would be good too. I think there's some precedent in Tomcat for this type of thing where it will log the first occurrence at WARN/INFO and subsequent occurrences at DEBUG.

Probably not to be done in this repo, but logging the configured reporters at construction of the `Tracing` object could help with this, while adding to startup logs in all applications. Alternatively, Sleuth could do some logging like this in one log line at startup with various information about what is configured.

**Comment by adriancole** [comment]
@shakuzen Thanks, these are good suggestions. putting on your micrometer hat. I completely forgot that in our finagle README, we show how to use their built-in metrics handler to redirect annotation metrics to logs. Is there anything like this in micrometer?

Here's the snippet:

### Programmatic

You can also configure zipkin tracers programmatically. Here's an example:

```java
HttpZipkinTracer.Config config = HttpZipkinTracer.Config.builder()
    // The frontend makes a sampling decision (via Trace.letTracerAndId) and propagates it downstream.
    // This property says sample 100% of traces.
    .initialSampleRate(1.0f)
    // All servers need to point to the same zipkin transport
    .host("127.0.0.1:9411").build();

Tracer tracer = HttpZipkinTracer.create(config,
    // print stats about zipkin to the console
    new JavaLoggerStatsReceiver(Logger.getAnonymousLogger()));
```

**Comment by shakuzen** [comment]
> I completely forgot that in our finagle README, we show how to use their built-in metrics handler to redirect annotation metrics to logs. Is there anything like this in micrometer?

In Micrometer, there's a LoggingMeterRegistry which will periodically output all registered Micrometer metrics to logs. In general, for production I would suggest an implementation with a metrics backend, but it can be useful for debugging locally or in a test environment.

**Comment by meltsufin** [comment]
It sounds like the main reason these severe errors are logged at the debug level is the attempt to avoid the flood of output in the log. However, it's semantically incorrect, at least for the first instance of the error. So, is there anything wrong with the ideas already mentioned here of using a logger that either rate-limits the output, or simply outputs at a higher level at first, and then drops to debug? Can this change be made in the `AsyncReporter`?

**Comment by adriancole** [comment]
I think people are assuming I personally should do all the work, hence it
not happening :) I'm so incredibly knackered cleaning up several projects
worth of tech debt.

please can someone just implement either of the options and not expect me
to both coordinate and perform all work on several repos simultaneosly??

easiest is the log once imho

**Comment by adriancole** [comment]
the easiest way to do log once is to use a volaite or atomic boolean here
and first time switch level to WARN with a suffix about this being a once
thing and look at debug level or metrics for more.

https://github.com/openzipkin/zipkin-reporter-java/blob/master/core/src/main/java/zipkin2/reporter/AsyncReporter.java#L288-L292

there are ways to test logger output to see if the message is at a level
needed. you can look around I think here or brave they exist.

the info at startup thing can be followed up on at sleuth. our tracers have
pretty toString now, so that might be enough.

please whoever does this do not add a core dependency on anything... just
address the log once thing. everyone always assumes slf4j 1.7 or whatever
version is constant and it isn't

**Comment by adriancole** [comment]
actually the flush thread is single threaded.. the boolean doesn't
even need to atomic or anything. should be a good first time issue

**Comment by dzou** [comment]
I'll take a look at this.
